README
= kiva -- wrapper to the kiva rest API

This is a wrapper to the Kiva Web API. You can find more information
about the API here[http://developers.wiki.kiva.org/KivaAPI].


== Example

The following is an example of retrieving the last 20 loans on Kiva.
This is about as difficult as things get:

    require 'kiva'
    
    newest = Kiva::Loan.load_newest
    newest.each{ |loan|
      puts "#{loan.name} loaned $#{loan.funded_amount} for
			#{loan.activity} in #{loan.location["country"]}"
    }

Running the above script yields something similar to the following:

    $ ruby -rubygems examples/newest.rb 

    Angela Okosun loaned $225 for Cement in Nigeria
    Vitaliy Konovalenko loaned $1650 for Retail in Ukraine
		(...)
    Thi Thoa Do's Group loaned $300 for Food Production/Sales in Viet Nam
    Thi Hoa Nguyen loaned $900 for General Store in Viet Nam


== Prerequisites

Kiva doesn't require an account or an API key in order to use their API,
so there are no prerequisites to using it. This library does depend on
SimpleHTTP[http://simplehttp.rubyforge.org/] and
json[http://json.rubyforge.org/] to handle the web api. These should be
installed automatically if you used +gem+ to install the library, else,
you'll need to install them manually.


== Basic Usage

This library consists of number of classes modelling the data
objects which the web api returns. Each class is similar, consisting of
class methods used to call webapi functions and attributes which provide
access to the (unpacked) json data the webapi returns.

These classes are (in no particular order) :

* <code>Kiva::Loan</code>
* <code>Kiva::Lender</code>
* <code>Kiva::LendingAction</code>
* <code>Kiva::Partner</code>
* <code>Kiva::JournalEntry</code>
* <code>Kiva::Comment</code>
* <code>Kiva::Release</code>

Each of these classes has attributes containing the data returned from
the web api. For example, a +Comment+ has the attributes:

* <code>author</code>
* <code>body</code>
* <code>date</code>
* <code>id</code>
* <code>whereabouts</code>

Each class has one or more class methods, corresponding to web api
calls, which may be used to create and load instances of the class. For
example, +Loan+ has the following list of class methods:

* +load+ : loans/&lt;ids&gt;[http://developers.wiki.kiva.org/KivaAPI#loans/ltidsgt]
* +load_for_lender+ : loans/&lt;id&gt;/lenders[http://developers.wiki.kiva.org/KivaAPI#loans/ltidgt/lenders]
* +load_newest+ : loans/newest[http://developers.wiki.kiva.org/KivaAPI#loans/newest]
* +search+ : loans/search[http://developers.wiki.kiva.org/KivaAPI#loans/search]



== Installing

You should be able to `gem install kiva`. Alternatively, download a distribution
at rubyforge: http://rubyforge.org/frs/?group_id=8034 

Finally you can grab the sources from the rubyforge or github  repositories:

git clone git://rubyforge.org/kiva.git

or

http://github.com/a2800276/kiva/tree/master#



== Mailing List

In case you discover bugs, spelling errors, offer suggestions for
improvements or would like to help out with the project, you can contact
me directly (tim@kuriositaet.de).
	


CHANGELOG
= Version 0.1.14593
=== Fri May 15 11:19:25 CEST 2009
`Filter` is extracted, there are two classes: LoanFilter and
JournalFilter, the previous `Filter` is now LoanFilter.

now handles the `search` method of JournalEntries

bugfix handling Comment loading

change minor version (14593) to correspond to kiva release number


= Version 0.1.0
=== Wed May 13 14:56:26 CEST 2009
Updated tests to check all expected attributes and urls are 
available from kiva

= Version 0.0.2
=== Mon Mar 23 16:51:54 CET 2009
* nicer documentation

= Version 0.0.1
=== Wed Mar 18 17:59:23 CET 2009
* initial version, hope everything works 

back[link:files/README.html]


test/test_kiva.rb
require 'test/unit'
require 'pp'
require File.dirname(__FILE__) + '/../lib/kiva'


module Kiva
  fixtures = File.open(File.dirname(__FILE__)+"/fixtures.rbf").readlines.join
  $fixtures = eval(fixtures)

	# Modified `execute` method for TESTING!
  # Patch up execute method to use dummies and not connect to real web api.
  def Kiva.execute url, query=nil
    key = [url, query]
    $fixtures[key]
  end
end

class TestKiva < Test::Unit::TestCase
  def test_load_lender
    lender = Kiva::Lender.load("tim9918")
    assert_equal(1, lender.length)
    lender = lender[0]
    assert_equal("DE", lender.country_code)
    assert_equal(3, lender.invitee_count)
    assert_equal("tim9918", lender.lender_id)
    assert_equal("", lender.loan_because)
    assert_equal(114, lender.loan_count)
    assert_equal("2007-04-09T16:37:58Z", lender.member_since)
    assert_equal("Tim", lender.name)
    assert_equal("Computer Programmer", lender.occupation)
    assert_equal("", lender.occupational_info)
    assert_equal("www.kuriositaet.de", lender.personal_url)
    assert_equal("tim9918", lender.uid)
    assert_equal("K\303\266ln", lender.whereabouts)
  end

  def test_load_lender_for_loan
    lenders = Kiva::Lender.load_for_loan 95693

    assert_equal(35, lenders.length)

    assert_equal("US", lenders[0].country_code)
    assert_equal("eleanor1399", lenders[0].lender_id)
    assert_equal("eleanor1399", lenders[0].uid)
    assert_equal("Cambridge MA", lenders[0].whereabouts)

    assert_equal("US", lenders[34].country_code)
    assert_equal("joeandchris1024", lenders[34].lender_id)
    assert_equal("joeandchris1024", lenders[34].uid)
    assert_equal("Joe and Chris", lenders[34].name)
    assert_equal("princeton NJ", lenders[34].whereabouts)

  end

  def test_load_loan
     loan = Kiva::Loan.load_for_lender("tim9918")
     
     assert_equal(20, loan.length)
    
     # Test the first and the last loan in the array.

     assert_equal("Retail", loan[0].activity)
     assert_equal({"languages"=>["ru", "en"]}, loan[0].description)
     assert_equal(2675, loan[0].funded_amount)
     assert_equal(95189, loan[0].id)
     assert_equal({"template_id"=>1, "id"=>288221}, loan[0].image)
     assert_equal(
      {"country"=>"Tajikistan",
       "geo"=>
              {"type"=>"point", 
               "level"=>"town", 
               "pairs"=>"40.116667 70.633333"},
       "town"=>"Isfara"}, loan[0].location)
     assert_equal("Saboat Artykova", loan[0].name)
     assert_equal(47, loan[0].partner_id)
     assert_equal(Time.parse("Thu Mar 19 02:10:07 UTC 2009"), loan[0].posted_date)
     assert_equal("Retail", loan[0].sector)
     assert_equal("funded", loan[0].status)
     assert_equal("To expand the business", loan[0].use)




     assert_equal("Electrical Goods", loan[19].activity)
     assert_equal({"languages"=>["en"]}, loan[19].description)
     assert_equal(425, loan[19].funded_amount)
     assert_equal(79959, loan[19].id)
     assert_equal({"template_id"=>1, "id"=>240444}, loan[19].image)
     assert_equal(
      {"country"=>"Nigeria",
       "geo"=>
              {"type"=>"point", 
               "level"=>"town", 
               "pairs"=>"6.333333 5.633333"},
       "town"=>"Benin City"}, loan[19].location)
     assert_equal("Victor Okonkwo", loan[19].name)
     assert_equal(20, loan[19].partner_id)
     assert_equal("Wed Dec 17 15:50:11 UTC 2008", loan[19].posted_date.to_s)
     assert_equal("Retail", loan[19].sector)
     assert_equal("in_repayment", loan[19].status)
     assert_equal("To purchase electrical materials for sell", loan[19].use)

  end

  def test_lending_action
    la = Kiva::LendingAction.load
    #pp la[99]
    assert_equal(100, la.length)

    #sample first and last action
    assert_equal(11661649, la[0].id)
    assert_equal("Ruth", la[0].lender.name)
    assert_equal("Agriculture", la[0].loan.sector)

    assert_equal(11661488, la[99].id)
    assert_equal("James McGovern", la[99].lender.name)
    assert_equal("Clothing", la[99].loan.sector)
  end

  def test_journal_entry_comment
    je = Kiva::JournalEntry.load 14077
    
    assert_equal(1, je.length)
    assert_equal("Luis Crespo", je[0].author)
    assert_equal(false, je[0].bulk)
    assert_equal(1, je[0].comment_count)
    
    comments = Kiva::Comment.load(je[0])
    
    assert_equal(1, comments.length)
    assert_equal(16958, comments[0].id)
    assert_equal("Janet and Marty", comments[0].author)
    assert_equal("Dear Dona Isabel,\r\n\r\n", comments[0].body)
    assert_equal(Time.parse("Wed Jul 18 16:52:53 UTC 2007"), comments[0].date)
    assert_equal("San Francisco, California, USA", comments[0].whereabouts)
  end

  def test_partner
    partner = Kiva::Partner.load
    assert_equal(109, partner.length)

    assert_equal(0, partner[0].default_rate)
    assert_equal(128, partner[0].id)
    assert_equal(99, partner[0].loans_posted)
    assert_equal("Hagdan sa Pag-uswag Foundation, Inc. (HSPFI)", partner[0].name)

    assert_equal(3, partner[108].countries.length)
    assert_equal(9.1917293233083, partner[108].default_rate)
    assert_equal(1, partner[108].id)
    assert_equal("closed", partner[108].status)

  end

  def test_template
    templates = Kiva::Templates.load
    assert_equal(1, templates.length)
    assert_equal(1, templates[0].id)
    assert_equal("http://www.kiva.org/img/<size>/<id>.jpg", templates[0].pattern)
  end

  def test_release
    release = Kiva::Release.load
    assert_equal("13775", release.id)
    assert_equal(Time.parse("Thu Mar 19 01:20:37 UTC 2009"), release.date)
  end
  
  #todo test loan search
  def test_journal_entry
    f = Kiva::JournalFilter.new.media_any
    entries = Kiva::JournalEntry.search f
    e = entries[0]
    assert_equal(20, entries.length)
    assert_equal(0, e.recommendation_count, "recommendation_count")
    assert_equal("<p>Check out this interview with Pauline as she discusses her hopes and challanges of her new loan!</p>", e.body, "body")
    assert_equal(false, e.bulk, "bulk")
    assert_equal("Kristy Callahan", e.author, "author")
    assert_equal(1, e.comment_count, "comment_count")
    assert_equal({"template_id"=>1, "id"=>318954}, e.image, "image")
    
    #assert_equal("", e.comments, "comments")
    comments = e.comments
    c = comments[0]
    assert_equal(1, comments.length)
    assert_equal("Kristy, thank you so much for sharing these interviews with the Kiva lenders.  It is so gratifying to hear their stories and see their faces.  ", c.body, "body")
    assert_equal("Suzanne Johannsen", c.author, "author")
    assert_equal("Bend, Oregon USA", c.whereabouts, "whereabouts")
    assert_equal("Thu May 14 03:06:11 UTC 2009", c.date.to_s, "date")
    
    

    assert_equal("Meet Pauline!", e.subject, "subject")
    assert_equal("Wed May 13 20:15:32 UTC 2009", e.date.to_s, "date")
    e = entries[0]
  end

  
end


test/generate_fixtures.rb
#!/usr/bin/env ruby -rubygems

require 'pp'
require File.dirname(__FILE__) + '/../lib/kiva'

module Kiva
	# Modified `execute` method to generate fixtures!
  def Kiva.execute url, query=nil
    result = SimpleHttp.get(url, query)
    key = [url, query]
    $fixtures[key] = result
    result
  end
end

$fixtures = {}

user = Kiva::Lender.load("tim9918")
loans = Kiva::Loan.load_for_lender(user[0])
Kiva::LendingAction.load
Kiva::Lender.load_for_loan 95693
Kiva::JournalEntry.load 14077
je = Kiva::JournalEntry.load 14077
Kiva::Comment.load(je[0])
Kiva::Partner.load
Kiva::Templates.load
filter = Kiva::LoanFilter.new.male.africa
Kiva::Loan.search filter
Kiva::Release.load
f = Kiva::JournalFilter.new.media_any
e = Kiva::JournalEntry.search f
e[0].comments
pp $fixtures


lib/kiva.rb
require 'json'
require 'simplehttp'

# The Kiva module is a namespace to contain all the classes and contains
# some utility functions to handle json and http.

module Kiva

  #
  # This is central location where the webservice query
  # is performed. You probably won't need to change this, but in case you
  # you need special error handling or some other behaviour, you can
  # modify everything in one place HERE.
  # 
  # For examples on how to do this, have a look at the files: `test/generate_fixtures.rb`
  # and `test/test_kiva.rb`.
  #
  # ===Error Handling
  #
  # currently the normal HttpExceptions are just passed
  # on, which should be ok for starters.
  #
  def Kiva.execute url, query=nil
    #puts url
    #pp(query) if query
    SimpleHttp.get(url, query)
  end
 
  #
  # INTERNAL
  #
  #   Utility to transfer JSON structs to 'real' classes.
  #
  #   Takes an instance of a class and a hash.
  #   for each key in hash, we check if the instance
  #   has a corresponding setter, if so, we assign the 
  #   value of the key to the instance.
  #
  # E.G.
  #
  #   If the hash is:
  #
  #   { :one => "bla", :two => "blo", :three => "blu"}
  #
  #   And an instance if of class:
  # 
  #     class Example
  #        attr_accessor :one
  #        attr_accessor :three
  #     end
  #
  #   `_fill` would set example.one="bla" and example.three="blu"
  #
  def Kiva._fill instance, hash
    hash.keys.each { |key|

      meth = (key+"=").to_sym
      next unless instance.respond_to? meth

      if ["loan", "lender"].include?(key) 
        value = ("loan"==key ? Kiva::Loan.new : Kiva::Lender.new)
        Kiva._fill value, hash[key]
      elsif key =~ /date/
        begin
          require 'time'
          value = Time.parse(hash[key], -1)
        rescue
          value = hash[key]
        end
      else
        value = hash[key]
      end

      instance.__send__ meth, value
    }
    return instance
  end

  #
  # INTERNAL
  #
  # Utility to transfer JSON structures to normal ruby arrays.
  #
  # array contains an array of parsed JSON structs, e.g.
  #
  #   [{"one"=>3},{"one"=>5},{"one"=>6}]
  #
  # given the name of a Class in the `clazz` param, this method
  # will instantiate one clazz per JSON struct in `array` and attempt
  # to populate the class. (see `_fill` above)
  def Kiva._populate clazz, array
    res = []
    array.each{ |hash|
      instance = clazz.new
      Kiva._fill instance, hash
      res.push instance 
    }
    res
  end

  # Represents a Loan (whoda guessed.) The accessible parameters are all
  # all there is to this. Depending on how a particular instance of loan
  # was loaded, not all of the attributes need have a value, so remember
  # to check for +nil+.
  #
  # Use one of the class functions:
  # * +load+
  # * +load_for_lender+
  # * +load_newest+
  # * +search+
  #
  # to load loans from Kiva.
  class Loan
    attr_accessor :id
    attr_accessor :status
    attr_accessor :name
    attr_accessor :posted_date
    attr_accessor :activity
    attr_accessor :borrowers
    attr_accessor :description
    attr_accessor :terms
    attr_accessor :partner_id
    attr_accessor :use
    attr_accessor :funded_amount
    attr_accessor :funded_date
    attr_accessor :image
    attr_accessor :location
    attr_accessor :sector
    attr_accessor :basket_amount

    attr_accessor :borrower_count
    attr_accessor :loan_amount

    
     
    KEY = "loans"
    
    LOAD_FOR_LENDER = "http://api.kivaws.org/v1/lenders/%s/loans.json?"
    LOAD_NEWEST     = "http://api.kivaws.org/v1/loans/newest.json?"
    LOAD            = "http://api.kivaws.org/v1/loans/%s.json"
    SEARCH          = "http://api.kivaws.org/v1/loans/search.json"
    
    class << self
      
      

      #
      # Returns details for one or more loans. 
      #
      # ====Paramaters
      # +ids+ : an instance of +Loan+ or a loan id or an array +Loan+'s
      # or loan id's
      #
      # ====Returns
      # an array of +Loan+ instances
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#loans/ltidsgt
      #
      def load ids
        case ids
        when Loan
          ids = ids.id
        when Array
          if ids[0].is_a?(Loan)
            ids = ids.map{|l| l.id}
          end
          ids.join(",")
        end

        url = LOAD % ids

        raw = Kiva.execute url
        unw = JSON.parse(raw)

        Kiva._populate  Loan, unw[KEY]
      end

      #
      # Search for loans matching specific criteria. 
      #
      # ====Parameters
      # +filter+ : an instance of +LoanFilter+ describing the search parameter 
      #
      # ====Returns
      # an array of +Loan+ instances
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#loans/search 
      #
      def search filter, page=nil
        url = SEARCH
        filter ||= LoanFilter.new
        if page
          filter["page"] = page
        end
        
        raw = Kiva.execute url, filter.params
        unw = JSON.parse(raw)

        Kiva._populate Loan, unw[KEY]

      end

      #
      # Returns the most recent fundraising loans. 
      #
      # ====Parameters
      # +page+ : the page position
      #
      # ====Returns
      # an array of +Loan+ instances
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#loans/newest
      #
      def load_newest page=nil
        url = LOAD_NEWEST
        url = page ? url + "page=#{page}&" : url
        
        raw = Kiva.execute url
        unw = JSON.parse(raw)

        Kiva._populate  Loan, unw[KEY]
      end
     

      # 
      # Returns loans sponsored by a specified lender
      # 
      # ====Parameters
      # * +id+ : id of lender or instance of +Lender+
      # * +sort_by+: one of <code>:newest, :oldest</code>
      # * +page+ : page position
      #
      # ====Returns
      # array of +Loan+'s
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#lenders/ltlenderidgt/loans
      #

      def load_for_lender id, sort_by=nil, page=nil
        id = id.uid if id.is_a?(Lender)
       
        url = LOAD_FOR_LENDER % id
        url = page ? url + "page=#{page}&" : url
        url = [:newest, :oldest].include?(sort_by) ? url + "sort_by=#{sort_by}" : url

        raw = Kiva.execute url
        unw = JSON.parse(raw)
        

        Kiva._populate  Loan, unw[KEY]
        
      end
    end

  end
  
  # Represents a Lender (whoda guessed.) The accessible parameters are all
  # all there is to this. Depending on how a particular instance of loan
  # was loaded, not all of the attributes need have a value, so remember
  # to check for +nil+.
  #
  # Use one of the class functions:
  # * +load+
  # * +load_for_loan+
  #
  # to load lender information from Kiva.

  class Lender
    # url = "http://api.kivaws.org/v1/lenders/%s.json"
    attr_accessor :uid
    attr_accessor :lender_id
    attr_accessor :name
    attr_accessor :loan_count
    attr_accessor :occupation
    attr_accessor :country_code
    attr_accessor :loan_because
    attr_accessor :invitee_count
    attr_accessor :occupational_info
    attr_accessor :personal_url
    attr_accessor :whereabouts
    attr_accessor :image
    attr_accessor :member_since
    attr_accessor :personal_url

    
    KEY  = "lenders"
      
    LOAD = "http://api.kivaws.org/v1/lenders/%s.json"
    LOAD_FOR_LOAN = "http://api.kivaws.org/v1/loans/%s/lenders.json?"
    class << self
     
      
      #
      # List public lenders of a loan.
      #
      # ====Parameters
      # +loan+ : a loan id or an instance of +Loan+ 
      #
      # ====Returns
      # an array of +Lender+ instances.
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#loans/ltidgt/lenders
      #
      def load_for_loan loan, page = nil
        loan = loan.id if loan.is_a?(Loan)

        url = LOAD_FOR_LOAN % loan

        url = page ? url + "page=#{page}&" : url
        
        raw  = Kiva.execute(url)
        unw = JSON.parse(raw)

        Kiva._populate Lender, unw[KEY]

      end

      #
      # Load details for one or more Lenders.
      #
      # ====Parameters
      # +ids+ : a lender id or and array of id's
      #
      # ====Returns
      # an array of +Lender+ instances.
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#lenders/ltlenderidsgt
      #
      def load ids
        ids  = ids.join(",") if ids.is_a?(Array)
        url  = LOAD % ids
        raw  = Kiva.execute(url)
        unw = JSON.parse(raw)

        Kiva._populate Lender, unw[KEY]
      end

    end

  end

  # +LendingAction+ are basically a showcase of recent Kiva activity.
  # Use the class method +load+ to loan the last 100 lenders and the
  # loan they sponsored.

  class LendingAction
    attr_accessor :id
    attr_accessor :date
    attr_accessor :lender
    attr_accessor :loan

    KEY = "lending_actions"
    LOAD_RECENT = "http://api.kivaws.org/v1/lending_actions/recent.json"

    class << self

      #
      # Returns the last 100 public actions from Kiva.
      #
      # ====Returns
      # an array of +LendingAction+ instances
      #
      # ====corresponds
      # http://developers.wiki.kiva.org/KivaAPI#lendingactions/recent
      #
      def load 
        raw = Kiva.execute(LOAD_RECENT)
        unw = JSON.parse(raw)
        
        Kiva._populate  LendingAction, unw[KEY]

      end
    end
  end
  
  # Journal entries are attached to loans. Using the class method +load+
  # you can retrieve all entries attached to a particular loan.
  #
  
  class JournalEntry
    attr_accessor :id
    attr_accessor :body
    attr_accessor :date
    attr_accessor :comment_count
    attr_accessor :author
    attr_accessor :subject
    attr_accessor :bulk
    attr_accessor :image
    attr_accessor :recommendation_count

    #
    # Retrieve an array of +Comment+ instances
    # associated with this entry. N.B. this leads to
    # a further network call.
    def comments
      unless @comments 
        return nil if id.nil?
        if @comment_count != 0
          @comments = Comment.load(self)
        else
          @commetns = []
        end
      end
      @comments
    end

    KEY    = "journal_entries"
    LOAD   = "http://api.kivaws.org/v1/loans/%s/journal_entries.json?"
    SEARCH = "http://api.kivaws.org/v1/journal_entries/search.json?"
    class << self

      #
      # Load journal entries for a loan. 
      #
      # ====Parameters
      # +id+ : a loan id or an instance of +Loan+
      #
      # ====Returns
      # an array of +JournalEntry+ instances.  
      #
      # ====Corresponds 
      # http://developers.wiki.kiva.org/KivaAPI#loans/ltidgt/journalentries
      #
      def load id, page = nil, include_bulk = nil
        id = id.id if id.is_a? Loan
        url = LOAD % id
        
        url = page ? url + "page=#{page}&" : url
        url = (! include_bulk.nil?) ? url + "include_bulk=#{include_bulk}" : url
        
        raw = raw = Kiva.execute(url)
        unw = JSON.parse(raw)
        Kiva._populate JournalEntry, unw[KEY]

      end 
      

       #
      # Search for journal entries. 
      #
      # ====Parameters
      # +filter+ : an instance of JournalFilter defining the search 
      # +sort_by+ : one of  [:newest, :oldest, :recommendation_count, :comment_count]
      # +page+ : page to load
      # ====Returns
      # an array of +JournalEntry+ instances.  
      #
      # ====Corresponds 
      # http://developers.wiki.kiva.org/KivaAPI#journalentries/search
      #
  
      def search filter, sort_by=nil, page=nil
        url = SEARCH
        filter ||= JournalFilter.new 
        if sort_by && [:newest, :oldest, :recommendation_count, :comment_count].include?(sort_by)
          filter["sort_by"]=sort_by
        end

        if page
          filter["page"]=page
        end
        
        raw  = Kiva.execute url, filter.params
        unw = JSON.parse(raw)

        Kiva._populate JournalEntry, unw[KEY]
      end
    end
  end
  
  # User comment made in response to a JournalEntry.
  class Comment
    # http://api.kivaws.org/v1/journal_entries/<id>/comments.json
    attr_accessor :body
    attr_accessor :date
    attr_accessor :author
    attr_accessor :id
    attr_accessor :whereabouts

    KEY = "comments"
    URL = "http://api.kivaws.org/v1/journal_entries/%s/comments.json?"
    class << self
      
      #
      # Loads an array of comments for a JournalEntry.
      #
      # ====Parameters
      # * +id+  : the numerical id of a Journal Entry or an instance of the class +JournalEntry+
      # * +page+: which page of comments to load, default is the first.
      #
      # ====Returns
      # array of +Comments+
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#journalentries/ltidgt/comments
      #
      def load id, page=nil
        id = id.id if id.is_a?(JournalEntry)
     
        url = URL % id
        url = page ? url + "page=#{page}&" : url

        raw = raw = Kiva.execute(url)
        unw = JSON.parse(raw)
        Kiva._populate Comment, unw[KEY]
      end
    end
  end

  # Kiva field Partner
  class Partner
    attr_accessor :start_date
    attr_accessor :rating
    attr_accessor :status
    attr_accessor :name
    attr_accessor :delinquency_rate
    attr_accessor :id
    attr_accessor :total_amount_raised
    attr_accessor :default_rate
    attr_accessor :loans_posted
    attr_accessor :countries
    attr_accessor :image
  
    KEY  = "partners"
    LOAD = "http://api.kivaws.org/v1/partners.json?"
    class << self
      
      #
      # Load an alphabetically sorted list of partners.
      #
      # ====Parameters
      # +page+: page position 
      #
      # ====Returns
      # an array of +Partner+ instances
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#partners 
      #

      def load page=nil
        url = page ? LOAD + "page=#{page}" : LOAD
        raw = raw = Kiva.execute(url)
        unw = JSON.parse(raw)
        
        Kiva._populate self, unw[KEY]
      end
    end
  end

  # Release information concerning the webapi. This is not the
  # release number of the ruby library!
  class Release

    attr_accessor :date
    attr_accessor :id

    class << self
      #
      # Returns release information 
      #
      # ====Returns
      #   an array of +Partner+ instances
      #
      # ====Corresponds
      # http://developers.wiki.kiva.org/KivaAPI#releases/api/current 
      #
     
      def load
        url = "http://api.kivaws.org/v1/releases/api/current.json"
        raw = JSON.parse(Kiva.execute(url))
        Kiva._fill(self.new, raw["release"])
      end
    end
  end

  # Templates which may be used to construct html image tags.
  class Templates
    attr_accessor :id
    attr_accessor :pattern
    class << self
      def load 
        url = "http://api.kivaws.org/v1/templates/images.json"
        unw = JSON.parse(Kiva.execute(url))
        Kiva._populate self, unw["templates"]
      end
    end
  end

  class Filter
    attr_accessor :params
    def initialize
      @params = {}
    end

    def []= key, value
      @params[key] = value
    end

    def [] key
      @params[key]
    end
  end

  # Filter to be used in order to describe the intended search results
  # of <code>Loan.search</code>
  class LoanFilter < Filter
    
    # sort_by
    def popularity
      @params["sort_by"]="popularity"
      return self
    end
    # sort_by
    def loan_amount
      @params["sort_by"]="loan_amount"
      return self
    end
    # sort_by
    def oldest
      @params["sort_by"]="oldest"
      return self
    end
    # sort_by
    def expiration
      @params["sort_by"]="expiration"
      return self
    end
    # sort_by
    def newest
      @params["sort_by"]="newest"
      return self
    end
    # sort_by
    def amount_remaining
      @params["sort_by"]="amount_remaining"
      return self
    end
    # sort_by
    def repayment_term
      @params["sort_by"]="repayment_term"
      return self
    end

    # restrict to male loan recipients
    def male
      @params["gender"]="male"
      return self
    end
    # restrict to female loan recipients
    def female
      @params["gender"]="female"
      return self
    end

    # restrict status
    def fundraising
      @params["status"]="fundraising"
      return self
    end
    # restrict status
    def funded
      @params["status"]="funded"
      return self
    end
    # restrict status
    def in_repayment
      @params["status"]="in_repayment"
      return self
    end
    # restrict status
    def paid
      @params["status"]="paid"
      return self
    end
    # restrict status
    def defaulted
      @params["status"]="defaulted"
      return self
    end
     
    #restrict region
    def north_america
      @params["region"]="na"
      return self
    end
    #restrict region
    def central_america
      @params["region"]="ca"
      return self
    end
    #restrict region
    def south_america
      @params["region"]="sa"
      return self
    end
    #restrict region
    def africa
      @params["region"]="af"
      return self
    end
    #restrict region
    def asia
      @params["region"]="as"
      return self
    end
    #restrict region
    def middle_east
      @params["region"]="me"
      return self
    end
    #restrict region
    def eastern_europe
      @params["region"]="ee"
      return self
    end


    def sector= sector
      @params["sector"]=sector
    end
    def country_code= iso_2
      @params["country_code"]=iso_2
    end
    
    #def partner
    #  too lazy
    #end

    def query= query
      @params["q"]=query
    end
    

  end

  class JournalFilter < Filter

    def media_video
      @params["media"] = "video"
      self
    end

    def media_image
      @params["media"] = "image"
      self
    end

    def media_any
      @params["media"] = "any"
      self
    end

    def include_bulk
      @params["include_bulk"] = "true"
      self
    end
    
    def no_include_bulk
      @params["include_bulk"] = "false"
      self
    end

    def partner partner
      # if partner is an instance of Partner, extract the id,
      # else assume (!) users are passing in a sensible id.
      partner = partner.is_a?(Partner) ? partner.id : partner

      # partners may be a list
      @partners ||= []
      @partners.push(partner.to_s.strip)
      @params["partner"] = @partners.join(",")
      self
    end
  end # JournalFilter

 


end

if $0 == __FILE__
  url = "http://api.kivaws.org/v1/partners.json?"
  userid = "14077"
#
#  result = SimpleHttp.get(url % userid)
#
#  #puts(result)
#
#  puts ('-----------')
#  
#  results = JSON.parse(result)
#  pp results
#
#
#  results["partners"][0].keys.each { |key|
#    puts "attr_accessor :#{key}"
#  }
  #user = Kiva::Lender.load("tim9918") 
  #pp user
  user = "tim9918"
  #loans = Kiva::Loan.load_for_lender(user)
  #pp loans
  #pp Kiva::LendingAction.load
  #pp Kiva::Lender.load_for_loan 95693
  #pp Kiva::JournalEntry.load 14077
  
  #je = Kiva::JournalEntry.load 14077
  #pp je

  #pp Kiva::Comment.load(je[0])

  #pp (Kiva::Partner.load 2).length
  #pp Kiva::Templates.load

  #filter = Kiva::Filter.new.male.africa
  #pp  (Kiva::Loan.search filter)
  #pp Kiva::Release.load
end
